pkg/instance/service.go

```diff
@@ -76,43 +80,267 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
 	}
 
 	deployment.Instances = append(deployment.Instances, instance)
-	err = s.validateDeployment(deployment)
+
+	_, err = s.validateNoCycles(deployment.Instances)
+	if err != nil {
+		return err
+	}
+
+	err = s.resolveParameters(deployment)
+	if err != nil {
+		return err
+	}
+
+	err = s.validateDeploymentParameters(deployment)
 	if err != nil {
 		return err
 	}
 
 	return s.instanceRepository.SaveInstance(instance)
 }
 
-func (s service) validateDeployment(deployment *model.Deployment) error {
-	stacks, err := s.collectStacks(deployment)
+func (s service) validateNoCycles(instances []*model.DeploymentInstance) (graph.Graph[string, *model.DeploymentInstance], error) {
+	g := graph.New(func(instance *model.DeploymentInstance) string {
+		return instance.StackName
+	}, graph.Directed(), graph.PreventCycles())
+
+	for _, instance := range instances {
+		err := g.AddVertex(instance)
+		if err != nil {
+			return nil, fmt.Errorf("failed adding vertex for instance %q: %v", instance.Name, err)
+		}
+	}
+
+	for _, src := range instances {
+		stack, err := s.stackService.Find(src.StackName)
+		if err != nil {
+			return nil, err
+		}
+		for _, requiredStack := range stack.Requires {
+			err := g.AddEdge(src.StackName, requiredStack.Name)
+			if err != nil {
+				if errors.Is(err, graph.ErrEdgeAlreadyExists) {
+					return nil, fmt.Errorf("instance %q requires %q more than once", src.Name, requiredStack.Name)
+				} else if errors.Is(err, graph.ErrEdgeCreatesCycle) {
+					return nil, fmt.Errorf("edge from instance %q to instance %q creates a cycle", src.Name, requiredStack.Name)
```

Comment:
Should we use the work `link` when talking to users instead of `edge`? Since we already established the word link.

```suggestion
					return nil, fmt.Errorf("link from instance %q to instance %q creates a cycle", src.Name, requiredStack.Name)
```

Also we are using `instance` in the message but some of the args are stack names.